### ISSUE TITLE
False positive in DiscardedNotificationCenterObserverRule

### ISSUE THREAD

**Issue by sammy-SC** [realm/SwiftLint#1525]
Following code is labelled with warning even thought the opaque observer is getting preserved. 

```
return nc.addObserver(forName: .NSSystemTimeZoneDidChange, object: nil, queue: nil, using: { })\n
```

**Comment by marcelofabri** [comment]
This could be hard to fix, since there's no guarantee that the places that call this method will retain the observer object.

**Comment by sammy-SC** [comment]
@marcelofabri I believe it will be sufficient to check whether the function returning observer doesn't have `@discardableResult` prefix. Swift compiler will do the rest for us giving `Result of call ...` warnings.

**Comment by marcelofabri** [comment]
That's a good idea! It should be possible, since we get the attribute from SourceKit:

**`file.swift`**
```swift
@discardableResult
func foo() -> Int{
    return 0
}
```

**`$ sourcekitten structure --file file.swift`**
```json
{
  "key.diagnostic_stage" : "source.diagnostic.stage.swift.parse",
  "key.substructure" : [
    {
      "key.offset" : 19,
      "key.nameoffset" : 24,
      "key.bodyoffset" : 37,
      "key.accessibility" : "source.lang.swift.accessibility.internal",
      "key.length" : 33,
      "key.name" : "foo()",
      "key.kind" : "source.lang.swift.decl.function.free",
      "key.bodylength" : 14,
      "key.attributes" : [
        {
          "key.attribute" : "source.decl.attribute.discardableResult"
        }
      ],
      "key.namelength" : 5
    }
  ],
  "key.offset" : 0,
  "key.length" : 53
}
```
